File: src/Common/TAC.cpp
#include "Common/TAC.h"


std::string to_string(TACOP op) {
	switch (op) {
		case TACOP::ADD: return "add";
		case TACOP::SUB: return "sub";
		case TACOP::MUL: return "mul";
		case TACOP::DIV: return "div";
		case TACOP::ASSIGN: return "assign";
		case TACOP::CALL: return "call";
		case TACOP::PARA: return "para";
		case TACOP::RET: return "ret";
		case TACOP::ENDCALL: return "endcall";
		case TACOP::FUN_RET: return "funret";
		case TACOP::FUN_PARA: return "funpara";
		case TACOP::IFEXP: return "ifexp";
		case TACOP::ENDIF: return "endif";
		case TACOP::GOTO: return "goto";
		case TACOP::ELSE: return "else";
		case TACOP::IFGT: return "ifgt";
		case TACOP::IFGE: return "ifge";
		case TACOP::IFLT: return "iflt";
		case TACOP::IFLE: return "ifle";
		case TACOP::IFEQ: return "ifeq";
		case TACOP::IFNEQ: return "ifneq";
		case TACOP::LABEL: return "label";
		default: return "error_op";
	}
}

File: go_compiler/common/tac.py
from enum import Enum, auto
from typing import TypeAlias
from .UseInfo import UseInfo
from .Scope import Scope


class TACOP(Enum):
    ADD = auto()
    SUB = auto()
    MUL = auto()
    DIV = auto()
    ASSIGN = auto()
    CALL = auto()
    PARA = auto()
    RET = auto()
    ENDCALL = auto()
    FUN_RET = auto()
    FUN_PARA = auto()
    GOTO = auto()
    IFGT = auto()
    IFGE = auto()
    IFLT = auto()
    IFLE = auto()
    IFEQ = auto()
    IFNEQ = auto()
    LABEL = auto()
    CREATLIST = auto()


tacopmap = {
    TACOP.IFEQ: "je",
    TACOP.IFGE: "jge",
    TACOP.IFGT: "jg",
    TACOP.IFLE: "jle",
    TACOP.IFLT: "jl",
    TACOP.IFNEQ: "jne",
}


def to_string(op: TACOP) -> str:
    return op.name.lower()


class TACOPERANDTYPE(Enum):
    IMM = auto()
    VAR = auto()
    NULL_ = auto()
    PTR = auto()
    LABEL = auto()  # eg: ELSE0


class Operand:
    def __init__(self, value: str, type_: TACOPERANDTYPE):
        self.value: str = value
        self.use_info: UseInfo = UseInfo()
        self.OperType: TACOPERANDTYPE = type_

    def __str__(self) -> str:
        return f"(v:{self.value}, ui:{self.use_info}, ot:{self.OperType})"


class TACLine:
    def __init__(
        self,
        line: int,
        op: TACOP,
        src1: Operand,
        src2: Operand,
        dst: Operand,
        scope: Scope,
    ):
        self.line: int = line
        self.op: TACOP = op
        self.src1: Operand = src1
        self.src2: Operand = src2
        self.dst: Operand = dst
        self.scope: Scope = scope

    def __str__(self):
        return f"{self.line}:(op:{to_string(self.op)}, src1:{self.src1}, src2:{self.src2}, dst:{self.dst})"

    def __repr__(self):
        return self.__str__()


TACBlock: TypeAlias = list[TACLine]
TACFile: TypeAlias = dict[str, TACBlock]
